src/dmd/expression.d

```diff
@@ -6527,7 +6527,9 @@ extern (C++) final class CondExp : BinExp
 
                     if (v.needsScopeDtor())
                     {
-                        if (!vcond)
+                        // a temporary is created for the condition
+                        // only if it's not an integral exp
+                        if (!vcond && !ce.econd.isIntegerExp())
```

Comment:
This is maybe a bit simplistic. The optimizer is pretty smarter, it is able to optimize array literal indexing, enums, const scalar variables and similar things.
The following will still segfault
```d
bool ct() { return 1+1 == 5; }
void main()
{
    enum arr = [ct()];
    auto a = arr[1*0] ? B().p : null;
}
struct B
{
    void* p;
    ~this() {}
}
```
Since we don't want tor run the optimizer here just yet you can just declare the `vcond` variable as const or immutable `bool` then the optimizer later on should be able to const fold it and eliminate the dead branch.
See https://github.com/dlang/dmd/blob/ce961799427fa9d022bd155313e724a794e222c1/src/dmd/optimize.d#L76-L81